theano/tensor/opt.py

```diff
@@ -1430,19 +1436,29 @@ def same_shape(self, x, y, dim_x=None, dim_y=None):
 
 class ShapeOptimizer(Optimizer):
     """Optimizer that serves to add ShapeFeature as an fgraph feature."""
-    def __init__(self):
-        Optimizer.__init__(self)
-
     def add_requirements(self, fgraph):
         fgraph.attach_feature(ShapeFeature())
 
     def apply(self, fgraph):
         pass
 
+
+class UnShapeOptimizer(Optimizer):
+    """Optimizer remove ShapeFeature as an fgraph feature."""
+    def apply(self, fgraph):
+        for feature in fgraph._features:
+            if isinstance(feature, ShapeFeature):
+                fgraph.remove_feature(feature)
```

Comment:
I agree, we could find a better solution. But currently this is the only way I know how we can make it work.

I don't find an alternative in a few minutes. I would go for making an issue about this problem. I don't think we should fix it here. If you have suggestion of alternative way of doing this write it.